tools/cmaketobzl/cmaketobzl.go

```diff
@@ -0,0 +1,359 @@
+/*
+ * Copyright 2019 The Kythe Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/kythe/llvmbzlgen/cmakelib/ast"
+	"github.com/kythe/llvmbzlgen/cmakelib/bindings"
+	"github.com/kythe/llvmbzlgen/writer"
+)
+
+type blockCounter struct {
+	begin string
+	end   string
+	count int
+}
+
+func NewCounter(begin string) *blockCounter {
+	return &blockCounter{begin, "end" + begin, 0}
+}
+
+func (bc *blockCounter) Count(text string) bool {
+	matched := true
+	if text == bc.begin {
+		bc.count += 1
+	} else if text == bc.end {
+		bc.count -= 1
+	} else {
+		matched = false
+	}
+	return matched || bc.count > 0
+}
+
+type options struct {
+	macroName   string
+	shouldPrint Predicate
+	shouldAdd   Predicate
+	excludePath Predicate
+}
+
+type eval struct {
+	p *ast.Parser
+	o options
+
+	w    *writer.StarlarkWriter
+	v    *bindings.Mapping
+	path []string
+}
+
+type Option func(*eval)
+type Predicate func(string) bool
+
+func PrintCommands(p Predicate) Option {
+	return func(e *eval) {
+		e.o.shouldPrint = p
+	}
+}
+
+func RecurseCommands(p Predicate) Option {
+	return func(e *eval) {
+		e.o.shouldAdd = p
+	}
+}
+
+func ExcludePaths(p Predicate) Option {
+	return func(e *eval) {
+		e.o.excludePath = p
+	}
+}
+
+func DefineVars(vars map[string]string) Option {
+	return func(e *eval) {
+		for k, v := range vars {
+			e.v.Set(k, v)
+		}
+	}
+}
+
+func Matching(pat string) Predicate {
+	return regexp.MustCompile(pat).MatchString
+}
+
+func NewEvaluator(w io.Writer, opts ...Option) *eval {
+	e := &eval{
+		p: ast.NewParser(),
+		w: writer.NewStarlarkWriter(w),
+		v: bindings.New(),
+		o: options{
+			macroName: "generated_cmake_targets",
+			shouldAdd: func(n string) bool { return n == "add_subdirectory" },
+		},
+	}
+	for _, o := range opts {
+		o(e)
+	}
+	return e
+}
+
+func (e *eval) parse(input io.Reader) (*ast.CMakeFile, error) {
+	return e.p.Parse(input)
+}
+
+func (e *eval) parseFile(path string) (*ast.CMakeFile, error) {
```

Comment:
Is this needed?

Reply:
"needed" no, but given that defer is function-scoped and I'd prefer to close the file handle sooner rather than later it seemed a good idea.